### PR TITLE
[4.0]  Add prepared statements for plg_authentication_cookie

### PR DESCRIPTION
**Pull request by HLeithner**
### Summary of Changes
Updated SQL queries to prepared statements and made some cleanups around the queries.

### Testing Instructions

**This PR affect the Joomla Authentication System - handle with care**

Use the plugin in all ways you can think of.

* Login
* Logout
* use remember me
* close/open browser
* try with and without force password reset

### Expected result
Nothing changed.

### CODE REVIEW

File: plugins/authentication/cookie/cookie.php [hunk 1]
```diff
@@ -129,7 +131,7 @@ public function onUserAuthenticate($credentials, $options, &$response)
 
 		// Find the matching record if it exists.
 		$query = $this->db->getQuery(true)
-			->select($this->db->quoteName(array('user_id', 'token', 'series', 'time')))
+			->select($this->db->quoteName(['user_id', 'token', 'series', 'time']))
 			->from($this->db->quoteName('#__user_keys'))
 			->where($this->db->quoteName('series') . ' = ' . $this->db->quote($series))
```
Comment (Quy) [inline plugins/authentication/cookie/cookie.php]: Bind.

File: plugins/authentication/cookie/cookie.php [hunk 2]
```diff
@@ -116,7 +117,8 @@ public function onUserAuthenticate($credentials, $options, &$response)
 		// Remove expired tokens
 		$query = $this->db->getQuery(true)
 			->delete('#__user_keys')
```
Comment (Quy) [inline plugins/authentication/cookie/cookie.php]: Apply `$db->quoteName`

File: plugins/authentication/cookie/cookie.php [hunk 3]
```diff
@@ -164,8 +168,9 @@ public function onUserAuthenticate($credentials, $options, &$response)
 			 * Delete all tokens for this user!
 			 */
 			$query = $this->db->getQuery(true)
-				->delete('#__user_keys')
-				->where($this->db->quoteName('user_id') . ' = ' . $this->db->quote($results[0]->user_id));
+				->delete($this->db->quoteName('#__user_keys'))
+				->where($this->db->quoteName('user_id') . ' = :userid')
+				->bind(':userid', $results[0]->user_id, ParameterType::INTEGER);
```
Comment (alikon) [inline plugins/authentication/cookie/cookie.php]: ` user_id varchar(150) NOT NULL,`
```suggestion
				->bind(':userid', $results[0]->user_id);
```

**Review by alikon — CHANGES_REQUESTED** [review]

File: plugins/authentication/cookie/cookie.php [hunk 4]
```diff
@@ -115,8 +116,9 @@ public function onUserAuthenticate($credentials, $options, &$response)
 
 		// Remove expired tokens
 		$query = $this->db->getQuery(true)
-			->delete('#__user_keys')
-			->where($this->db->quoteName('time') . ' < ' . $this->db->quote(time()));
+			->delete($this->db->quoteName('#__user_keys'))
+			->where($this->db->quoteName('time') . ' < :now')
+			->bind(':now', time(), ParameterType::INTEGER);
```
Comment (alikon) [inline plugins/authentication/cookie/cookie.php]: ` 'time' varchar(200) NOT NULL,`
```suggestion
			->bind(':now', time());
```

File: plugins/authentication/cookie/cookie.php [hunk 5]
```diff
@@ -16,6 +16,7 @@
 use Joomla\CMS\Plugin\CMSPlugin;
 use Joomla\CMS\User\User;
 use Joomla\CMS\User\UserHelper;
+use Joomla\Database\ParameterType;
```
Comment (Quy) [inline plugins/authentication/cookie/cookie.php]:
```suggestion
```

File: plugins/authentication/cookie/cookie.php [hunk 6]
```diff
@@ -115,8 +115,9 @@ public function onUserAuthenticate($credentials, $options, &$response)
 
 		// Remove expired tokens
 		$query = $this->db->getQuery(true)
-			->delete('#__user_keys')
-			->where($this->db->quoteName('time') . ' < ' . $this->db->quote(time()));
+			->delete($this->db->quoteName('#__user_keys'))
+			->where($this->db->quoteName('time') . ' < :now')
+			->bind(':now', time());
```
Comment (SharkyKZ) [inline plugins/authentication/cookie/cookie.php]: >Notice: Only variables should be passed by reference

Assign `time()` to a variable.

Comment (HLeithner) [reply plugins/authentication/cookie/cookie.php]: would sql NOW() work?

Comment (SharkyKZ) [reply plugins/authentication/cookie/cookie.php]: No, it's not the same.

Comment (HLeithner) [reply plugins/authentication/cookie/cookie.php]: ok

**Comment by SharkyKZ** [comment]
I have tested this item :white_check_mark: successfully on 494697c99aad24873f49c54f13530b91559ab28c<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25048">issues.joomla.org/tracker/joomla-cms/25048</a>.</sub>

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 494697c99aad24873f49c54f13530b91559ab28c<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25048">issues.joomla.org/tracker/joomla-cms/25048</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25048">issues.joomla.org/tracker/joomla-cms/25048</a>.</sub>

**Comment by wilsonge** [comment]
Thanks!

**Pull request merged**